routers/web/repo/fork_to_edit.go

```diff
@@ -0,0 +1,241 @@
+		if exist, err := git_model.IsBranchExist(ctx, editRepo.ID, branchName); err == nil && exist {
+			ctx.Data["Err_NewBranchName"] = true
+			ctx.RenderWithErr(ctx.Tr("repo.editor.branch_already_exists", branchName), tpl, form)
+			return false
+		}
+	}
+
+	// Check for protected branch
+	canCommitToBranch, err := context.CanCommitToBranch(ctx, ctx.Doer, editRepo, branchName)
+	if err != nil {
+		log.Error("CanCommitToBranch: %v", err)
+	}
+	if !canCommitToBranch.CanCommitToBranch {
+		ctx.Data["Err_NewBranchName"] = true
+		ctx.RenderWithErr(ctx.Tr("repo.editor.cannot_commit_to_protected_branch", branchName), tpl, form)
+		return false
+	}
+
+	return true
+}
+
+// pushToEditRepositoryOrError pushes the branch that editing will be applied on top of
+// to the user fork, if needed. On failure, it displays and returns an error. The
+// branch name to be used for editing is returned.
+func pushToEditRepositoryOrError(ctx *context.Context, editRepo *repo_model.Repository, branchName string, tpl templates.TplName, form any) (string, error) {
+	// If editing the same repository, no need to push anything
+	if editRepo.ID == ctx.Repo.Repository.ID {
+		return ctx.Repo.BranchName, nil
+	}
+
+	// If editing a user fork, first push the branch to that repository
+	baseRepo := ctx.Repo.Repository
+	baseBranchName := ctx.Repo.BranchName
+
+	log.Trace("pushBranchToUserRepo: pushing branch to user repo for editing: %s:%s %s:%s", baseRepo.FullName(), baseBranchName, editRepo.FullName(), branchName)
+
+	if err := git.Push(ctx, baseRepo.RepoPath(), git.PushOptions{
+		Remote: editRepo.RepoPath(),
+		Branch: baseBranchName + ":" + branchName,
+		Env:    repo_module.PushingEnvironment(ctx.Doer, editRepo),
+	}); err != nil {
+		ctx.RenderWithErr(ctx.Tr("repo.editor.fork_failed_to_push_branch", branchName), tpl, form)
+		return "", err
+	}
+
+	return branchName, nil
+}
+
+// updateEditRepositoryIsEmpty updates the the edit repository to mark it as no longer empty
+func updateEditRepositoryIsEmpty(ctx *context.Context, editRepo *repo_model.Repository) {
+	if !editRepo.IsEmpty {
+		return
+	}
+
+	editGitRepo, err := gitrepo.OpenRepository(git.DefaultContext, editRepo)
+	if err != nil {
+		log.Error("gitrepo.OpenRepository: %v", err)
+		return
+	}
+	if editGitRepo == nil {
+		return
+	}
+
+	if isEmpty, err := editGitRepo.IsEmpty(); err == nil && !isEmpty {
+		_ = repo_model.UpdateRepositoryCols(ctx, &repo_model.Repository{ID: editRepo.ID, IsEmpty: false}, "is_empty")
+	}
+	editGitRepo.Close()
+}
+
+func forkToEditFileCommon(ctx *context.Context, editOperation, treePath string, notEditableMessage any) {
+	// Check if the filename (and additional path) is specified in the querystring
+	// (filename is a misnomer, but kept for compatibility with GitHub)
+	filePath, _ := path.Split(ctx.Req.URL.Query().Get("filename"))
+	filePath = strings.Trim(filePath, "/")
+	treeNames, treePaths := getParentTreeFields(path.Join(ctx.Repo.TreePath, filePath))
```

Comment:
Why these two could be joint?

Reply:
I copied this from `editFile`, but it doesn't seem to be needed here so will simplify to `ctx.Repo.TreePath`.